ocaml/xapi/xapi_db_upgrade.ml

```diff
@@ -230,64 +235,53 @@ let upgrade_wlb_configuration =
-    	{- For control domains, we respect the administrator's choice of target:
-    	{ol
-    		{- t.dynamic_max := s.target}
-    		{- t.dynamic_min := s.target}}}
-    }
-*)
 let upgrade_vm_memory_for_dmc =
   {
-    description= "Upgrading VM memory fields for DMC"
-  ; version= (fun x -> x <= george)
+    description= "Upgrading VM memory fields to disable DMC"
+  ; version= (fun x -> x <= yangtze)
   ; fn=
       (fun ~__context ->
         debug
           "Upgrading VM.memory_dynamic_{min,max} in guest and control domains." ;
         let module VMC = Vm_memory_constraints.Vm_memory_constraints in
         let update_vm (vm_ref, vm_rec) =
-          if vm_rec.API.vM_is_control_domain then (
-            let target = vm_rec.API.vM_memory_target in
-            debug "VM %s (%s) dynamic_{min,max} <- %Ld" vm_rec.API.vM_uuid
-              vm_rec.API.vM_name_label target ;
-            Db.VM.set_memory_dynamic_min ~__context ~self:vm_ref ~value:target ;
-            Db.VM.set_memory_dynamic_max ~__context ~self:vm_ref ~value:target
-          ) else (* Note this will also transform templates *)
-            let safe_constraints =
-              VMC.reset_to_safe_defaults
-                ~constraints:
-                  {
-                    VMC.static_min= vm_rec.API.vM_memory_static_min
-                  ; dynamic_min= vm_rec.API.vM_memory_dynamic_min
-                  ; target= vm_rec.API.vM_memory_target
-                  ; dynamic_max= vm_rec.API.vM_memory_dynamic_max
-                  ; static_max= vm_rec.API.vM_memory_static_max
-                  }
-            in
-            debug "VM %s (%s) dynamic_{min,max},target <- %Ld"
-              vm_rec.API.vM_uuid vm_rec.API.vM_name_label
-              safe_constraints.VMC.static_max ;
-            Db.VM.set_memory_static_min ~__context ~self:vm_ref
-              ~value:safe_constraints.VMC.static_min ;
-            Db.VM.set_memory_dynamic_min ~__context ~self:vm_ref
-              ~value:safe_constraints.VMC.dynamic_min ;
-            Db.VM.set_memory_target ~__context ~self:vm_ref
-              ~value:safe_constraints.VMC.target ;
-            Db.VM.set_memory_dynamic_max ~__context ~self:vm_ref
-              ~value:safe_constraints.VMC.dynamic_max ;
-            Db.VM.set_memory_static_max ~__context ~self:vm_ref
-              ~value:safe_constraints.VMC.static_max
+          match
+            (vm_rec.API.vM_is_control_domain, vm_rec.API.vM_power_state)
+          with
+          | true, _ ->
+              let target = vm_rec.API.vM_memory_target in
+              debug "VM %s (%s) dynamic_{min,max} <- %Ld" vm_rec.API.vM_uuid
+                vm_rec.API.vM_name_label target ;
+              Db.VM.set_memory_dynamic_min ~__context ~self:vm_ref ~value:target ;
+              Db.VM.set_memory_dynamic_max ~__context ~self:vm_ref ~value:target
+          | false, `Halted ->
+              (* Note this will also transform templates *)
```

Comment:
Can we use VMC.set below rather than setting individual fields? Or even just replace this whole thing with a call to `Xapi_vm_helpers.enforce_memory_constraints` to reduce duplication?